plotter/sankey.go

```diff
@@ -0,0 +1,454 @@
+	// modified for individual stocks and flows.
+	color.Color
+
+	// StockBarWidth is the widths of the bars representing
+	// the stocks. The default value is 15% larger than the
+	// height of the stock label text.
+	StockBarWidth vg.Length
+
+	// LineStyle specifies the default border
+	// line style for the stocks and flows. Styles can be
+	// modified for individual stocks and flows.
+	draw.LineStyle
+
+	// TextStyle specifies the default stock label
+	// text style. Styles can be modified for
+	// individual stocks.
+	draw.TextStyle
+
+	flows []*Flow
+
+	// FlowStyle is a function that specifies the
+	// background colar and border line style of the
+	// flow based on its group name. The default
+	// function uses the default Color and LineStyle
+	// specified above for all groups.
+	FlowStyle func(group string) (color.Color, draw.LineStyle)
+
+	// stocks arranges the stocks by category.
+	// The first key is the category and the seond
+	// key is the label.
+	stocks map[int]map[string]*stock
+}
+
+// stock represents the amount of a stock and its plotting order.
+type stock struct {
+	receptorValue, sourceValue float64
+	label                      string
+	category                   int
+	order                      int
+
+	// min represents the beginning of the plotting location
+	// on the value axis.
+	min float64
+
+	// max is min plus the larger of receptorValue and sourceValue.
+	max float64
+
+	// sourceFlowPlaceholder and receptorFlowPlaceholder track
+	//  the current plotting location during
+	// the plotting process.
+	sourceFlowPlaceholder, receptorFlowPlaceholder float64
+}
+
+// A Flow represents the amount of an entity flowing between two stocks.
+type Flow struct {
+	// SourceStockLabel and ReceptorStockLabel are the labels
+	// of the stocks that originate and receive the flow,
+	// respectively.
+	SourceStockLabel, ReceptorStockLabel string
+
+	// SourceStockCategory and ReceptorStockCategory define
+	// the locations on the category axis of the stocks that
+	// originate and receive the flow, respectively. The
+	// SourceStockCategory must be a lower number than
+	// the ReceptorStockCategory.
+	SourceStockCategory, ReceptorStockCategory int
+
+	// Value represents the magnitute of the flow.
+	// It must be greater than or equal to zero.
+	Value float64
+
+	// Group specifies the group that a flow belongs
+	// to. It is used in assigning styles to groups
+	// and creating legends. If Group is blank,
+	// the flow will be assigned to the group "Default".
+	Group string
+
+	// inUse  is used to ensure that this
+	// Flow is only used in one Plotter.
+	inUse bool
+}
+
+// NewSankey creates a new Sankey diagram with the specified
+// flows and stocks.
+func NewSankey(flows ...*Flow) (*Sankey, error) {
+	s := new(Sankey)
+
+	s.stocks = make(map[int]map[string]*stock)
+
+	s.flows = flows
+	for i, f := range flows {
+		if f.inUse {
+			panic("this Flow is already in use in another Plotter")
```

Comment:
prepend error message with `"plotter:"`

also, perhaps replace with: `"plotter: Flow already in use in another Sankey diagram"`